### DISCUSSION TITLE
"Pretty" formatting for corfu-popupinfo

### DISCUSSION THREAD

**Discussion by aguynamedben** [minad/corfu#438] [Q&A] [answered]
I'm comparing different formatting levels of documentation and am wondering if there's a way to improve corfu-popupinfo's documentation preview (corfu-popupinfo) prettier. This is in Emacs 30.0.50.

### corfu-popupinfo-mode
There no syntax highlighting or horizontal rule.
<img width="851" alt="image" src="https://github.com/minad/corfu/assets/107841/62509308-93ed-44a2-a3f5-f831cc1a75fb">

### the stock eldoc buffer
Nice syntax highlighting and horizontal rule, but not displaying &nbsp; (from Markdown?) well.
<img width="851" alt="image" src="https://github.com/minad/corfu/assets/107841/71a61186-b544-4e9b-bcb0-fe2fcfadbc57">

### eldoc-box
Even nicer. eldoc-box README mentions they do some extra pretty formatting. Maybe it's easy to share code or gain inspiration from that project?
<img width="602" alt="image" src="https://github.com/minad/corfu/assets/107841/6ed65b01-a8fe-4732-b2f7-9714ef595ef9">

### company-quickhelp
Not so great either.
<img width="789" alt="image" src="https://github.com/minad/corfu/assets/107841/746cc238-665e-4eac-af1f-7eb4248a2f8b">

Is there a way to get prettier formatting in corfu-popupinfo?

**Comment by minad** [comment]
> Is there a way to get prettier formatting in corfu-popupinfo?

The content of the popup is provided in formatted form by the backend. Therefore it looks almost the same for corfu-popupinfo and company-quickhelp.

**Reply by aguynamedben** [reply]
Thanks for the discussion.

I'm looking closer at the eglot-events-buffer for a pyright session and a gopls session.

<img width="844" alt="image" src="https://github.com/minad/corfu/assets/107841/45de45c9-2856-4fc2-ae90-81ec76549e28">

<img width="847" alt="image" src="https://github.com/minad/corfu/assets/107841/fcc6ce32-70ae-4897-98b1-bbbe6430bc16">

In eldoc-box, it seems like he's taking the eldoc content and then running it through 6 pretty simple "prettify" functions that assume plaintext or Markdown. I think this is how he is achieving the nice look with "post processing" of what the LSP gives to make it look good in the UI. I think "making it pretty" vs. "just show whatever the language server gave" might be a worthy trade-off.
https://github.com/casouri/eldoc-box/blob/master/eldoc-box.el#L197

What don't understand is how the built-in eldoc buffer (along with eldoc-box) is doing syntax highlighting. It seems like eldoc might have it's own post-LSP-processing logic, note how the stock eldoc buffer further-highlights params and types:
<img width="851" alt="image" src="https://github.com/minad/corfu/assets/107841/541685b3-2c32-4da6-a3a3-276c4d4cab58">

I guess this is maybe a feature request for corfu-popupinfo extension. Somehow eldoc + eldoc-box are post-processing the LSP output and making it Very Nice™ looking.

**Reply by aguynamedben** [reply]
It seems that pyright's jsonrpc response for `completionItem/resolve` messages says `"kind":"plaintext"` but if you look at the content of the message it seems to be Python-docstring-formatting.

I think the cleanup is happening at the eglot level by markdown-mode

https://www.gnu.org/software/emacs/manual/html_node/eglot/Eglot-Features.html#

> If the popular third-party package markdown-mode is installed, and the server provides at-point documentation formatted as Markdown in addition to plain text, Eglot arranges for the ElDoc package to enrich this text with fontifications and other nice formatting before displaying it to the user. This makes the documentation shown by ElDoc look nicer on display.

Perplexity AI summarizes it as this:
> The integration between Eldoc and `markdown-mode` does not happen directly within the `eldoc.el` source code. Instead, the integration is facilitated by the presence of `markdown-mode` and the way Emacs' LSP client, such as Eglot, handles the documentation provided by language servers.
> 
> According to the Reddit post[1], if `markdown-mode` is installed and the language server provides documentation formatted as Markdown, Eglot will arrange for Eldoc to enrich the text with fontifications and other formatting before displaying it. This means that the actual integration is happening at the level of Eglot's handling of the LSP messages, rather than within Eldoc itself.
> 
> Eglot, as an Emacs LSP client, is designed to work with language servers and handle the various capabilities they provide, including documentation. When Eglot receives documentation in Markdown format from a language server, it uses `markdown-mode` to render this documentation with the appropriate formatting if `markdown-mode` is available.
> 
> The actual code that performs this integration would be part of Eglot's implementation, which would check for the presence of `markdown-mode` and use it to format the documentation strings before they are passed to Eldoc for display. This is why you do not see direct references to `markdown-mode` in the `eldoc.el` source code; the integration is abstracted away by the LSP client (Eglot in this case).
> 
> To summarize, the integration between Eldoc and `markdown-mode` for displaying formatted documentation from LSP servers is managed by the LSP client (like Eglot) and not by Eldoc itself. Eldoc simply displays the documentation string it receives, which has already been processed and formatted by the LSP client using `markdown-mode` if available[1].
>
> Citations:
> [1] https://www.reddit.com/r/emacs/comments/117nyw1/correctly_show_markdown_in_eldoc_buffer/
> [2] https://leanpub.com/markdown-mode/read
> [3] https://jblevins.org/projects/markdown-mode/
> [4] https://github.com/rust-lang/rust-analyzer/issues/3415
> [5] https://blog.sumtypeofway.com/posts/emacs-config.html
> [6] https://www.emacswiki.org/emacs/ElDoc
> [7] https://www.karsdorp.io/posts/dotemacs/
> [8] https://github.com/gabrielelana/dotfiles/blob/master/emacs.d/init.el

It seems this problem (and cleanup by eglot, eldoc, eldoc-box) might be particularly accute with PyRight. I do see some simple markdown with gopls which "properly" returns Markdown.

<img width="2161" alt="image" src="https://github.com/minad/corfu/assets/107841/99082b3a-6e55-463d-a45e-adf97a93b459">

I'll keep learning and keep an eye on it and let you know if I come across any ideas or ways to make the display better... thank you!

**Reply by aguynamedben** [reply, marked as answer]
Yeah, you can close this discussion. I think most of this nastiness was coming from PyRight, which seems to give back "nasty Python docstrings via plaintext" as it's docs.

I installed pylsp (https://github.com/python-lsp/python-lsp-server) and it's giving Markdown with corfu-popupinfo-mode is handling beautifully.

### From eglot-events-buffer
<img width="852" alt="image" src="https://github.com/minad/corfu/assets/107841/268be844-6e20-47b6-8e9d-de09b054dbad">

### corfu-popupinfo-mode
<img width="751" alt="image" src="https://github.com/minad/corfu/assets/107841/d661744b-5359-4922-b38b-a803d7b2fbd5">

Thank you for maintaining and working on Corfu, I love it so much!!! 🙏
